### PR TITLE
DataFrameIterator redesign

### PR DESCRIPTION
**Pull request by rragundez**
Objective

Change the design of the `DataFrameIterator` class to be more intuitive and flexible such that the `flow_from_dataframe` method of the `ImageDataGenerator` class handles:

- regression,
- binary classification,
- multi-class classification,
- multi-label classification,
- object detection,
- segmentation,
- Autoencoder,
- image noise removal,
- upscaling image to a higher resolution,
- custom semi-supervised learning use case,
- multi-task learning using any combination of the above use cases,
- multi-input with any of the above use cases,
- moonshot: multi-input where each input can have different data augmentation,

where the input data type is image data.

### CODE REVIEW

**Comment by rragundez** [comment]
Any updates? I also send the email to keras-users@googlegroups.com as requested.

**Comment by fchollet** [comment]
Hi there,

`DataframeIterator` will be deprecated in the future and replaced with a [redesigned image preprocessing API](https://github.com/keras-team/governance/blob/master/rfcs/20190729-keras-preprocessing-redesign.md). Have you checked it out?

**Comment by rragundez** [comment]
Hello thanks for the response. Yes, I have reviewed the redesign proposal quite carefully actually, and I saw the `from_dataframe` method. So that's where my proposal fits in, I believe I do mention it in my Redesign document. Currently that method is called `flow_from_dataframe` which instantiates a `DaraFrameIerator`, so I don't know if you are going to think of the application of `from_dataframe` in the same way, that is, instantiating a class, or just a function. In any case the ideas that I present here are applicable to the new design (as I mention in the document if I recall).

thanks again, let me know what you think as I would love to help out in this re-design of `from_dataframe`, as I think has great potential.

If you need me to, I can change the whole re-design proposal thinking in terms of the `from_dataframe` method, but it would mostly be renaming.

**Comment by rragundez** [comment]
@fchollet @Dref360 any update?

**Comment by fchollet** [comment]
Sorry for the late review. At this time we recommend using tf.data Dataset objects over Python generators for best performance, and as a result we are moving away from `ImageDataGenerator` and instead recommend the equivalent `image_dataset_from_directory` utility: https://keras.io/api/preprocessing/image/

It does not yet have dataframe support. We believe it would be valuable to add a `image_dataset_from_dataframe` utility, with a similar implementation.

Closing this PR. If are still interested in this feature, we can consider a `image_dataset_from_dataframe` PR.

**Comment by rragundez** [comment]
Thanks @fchollet , yeah a bit late the review but no worries. I will take a closer look at your comment and think about working on an implementation of `image_dataset_from_dataframe `.
